### PR TITLE
fix import a connected synapse workspace

### LINKED ISSUE #9845: Unable to attach synapse workspace to existing SQL pool
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

Terraform v0.14.2
hashicorp/azurerm v2.40.0

### Affected Resource(s)

* `azurerm_synapse_workspace`

### Terraform Configuration Files

```hcl
data "azuread_users" "adusers1" {
  user_principal_names = [ "example1@example.com", "example2@example.com" ]
}

data "azurerm_client_config" "current" {
}

resource "azurerm_resource_group" "rg" {
  name     = "exampleSynapse"
  location = "eastus2"
}

resource "azuread_group" "adgroup1" {
  name    = "synapseGroup"
  members = data.azuread_users.adusers1.object_ids
}

resource "azurerm_mssql_server" "sqlserver1" {
  name                          = "example-Synapse-SQL"
  location                      = azurerm_resource_group.rg.location
  resource_group_name           = azurerm_resource_group.rg.name
  version                       = "12.0"
  administrator_login           = "sqladminuser"
  administrator_login_password  = "example1234@"
  public_network_access_enabled = true
  minimum_tls_version           = "1.2"

  azuread_administrator {
    login_username = azuread_group.adgroup1.name
    object_id      = azuread_group.adgroup1.id
    tenant_id      = data.azurerm_client_config.current.tenant_id
  } 
}

resource "azurerm_mssql_database" "sqldatabase1" {
  name           = "exampleReporting"
  sku_name       = "DW100c"
  server_id      = azurerm_mssql_server.sqlserver1.id
  read_scale     = false
  zone_redundant = false
  collation      = "SQL_Latin1_General_CP1_CI_AS"
  create_mode    = "Default"
}

resource "azurerm_storage_account" "storage1" {
  name                     = "exampleSynapseDataLake"
  is_hns_enabled           = true
  account_tier             = "Standard"
  account_replication_type = "LRS"
  access_tier              = "Hot"
  account_kind             = "StorageV2"
  location                 = azurerm_resource_group.rg.location
  resource_group_name      = azurerm_resource_group.rg.name
  min_tls_version          = "TLS1_2"
}

resource "azurerm_role_assignment" "ra" {
  scope                = azurerm_storage_account.storage1.id 
  role_definition_name = "Storage Blob Data Contributor"
  principal_id         = azuread_group.adgroup1.id
}

resource "azurerm_storage_data_lake_gen2_filesystem" "filesystem1" {
  name               = "exampleSynapseDataLakeFileSystem"
  storage_account_id = azurerm_storage_account.storage1.id
}

resource "azurerm_synapse_workspace" "workspace1" {
  name                                 = "example-Synapse-SQL"
  resource_group_name                  = azurerm_resource_group.rg.name
  location                             = azurerm_resource_group.rg.location
  storage_data_lake_gen2_filesystem_id = azurerm_storage_data_lake_gen2_filesystem.filesystem1.id
  sql_administrator_login              = "sqladminuser"
  sql_administrator_login_password     = "example1234@"
}

data "azurerm_synapse_workspace" "dworkspace1" {
  name                = "example-Synapse-SQL"
  resource_group_name = azurerm_resource_group.rg.name
}

output "id" {
  value = data.azurerm_synapse_workspace.dworkspace1.id
} 
```

### Expected Behaviour

It is possible to create a Dedicated SQL pool (formerly SQL DW) with the azurerm_mssql_server and azurerm_mssql_database (with a DW*c sku) resources and then, via the portal, create a workspace associated with that SQL pool.

It is also possible to use hyphens in the workspace name.

It would be useful, if it is doable, for the azurerm_synapse_workspace resource to be used to associate an existing Dedicated SQL pool (formerly SQL DW) with the workspace and allow hyphens in the name.

### Actual Behaviour

The azurerm_synapse_workspace resource does not have an option to assign an existing Dedicated SQL pool (formerly SQL DW).

Attempt to import the workspace resource created through the portal throws:

```
{"code":"UnsupportedOperation","message":"For the workspace example-Synapse-SQL, this is an unsupported operation as it is a Connected workspace."}
```

Which appears to occur when running:

```
GET /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/exampleSynapse/providers/Microsoft.Synapse/workspaces/example-Synapse-SQL/administrators/activeDirectory?api-version=2019-06-01-preview HTTP/1.1
```

Attempt to retrieve the data source throws:

```
Error: name can contain only lowercase letters or numbers, and be between 1 and 45 characters long

  on main.tf line 324, in data "azurerm_synapse_workspace" "dworkspace1":
 324: data "azurerm_synapse_workspace" "dworkspace1" {
```

### Steps to Reproduce

1. `terraform apply`

### References

* #0000

### PR DESCRIPTION
**Pull request by ms-henglu**
A synapse workspace can be created from a dedicated SQL pool(formly SQL DW). And it can not perform GET aad admin/ sql aad admin operations.

To address: https://github.com/hashicorp/terraform-provider-azurerm/issues/9845

Fixes #9845

### CODE REVIEW

File: internal/services/synapse/synapse_workspace_resource.go [hunk 1]
```diff
@@ -477,13 +477,13 @@ func resourceSynapseWorkspaceRead(d *pluginsdk.ResourceData, meta interface{}) e
...
 	if err != nil {
-		if !utils.ResponseWasNotFound(aadAdmin.Response) {
+		if !utils.ResponseWasNotFound(aadAdmin.Response) && !utils.ResponseWasConflict(aadAdmin.Response) {
 			return fmt.Errorf("retrieving AzureAD Admin for %s: %+v", *id, err)
 		}
 	}
 	sqlAdmin, err := sqlAdminClient.Get(ctx, id.ResourceGroup, id.Name)
 	if err != nil {
-		if !utils.ResponseWasNotFound(sqlAdmin.Response) {
+		if !utils.ResponseWasNotFound(sqlAdmin.Response) && !utils.ResponseWasConflict(sqlAdmin.Response) {
```
Comment (tombuildsstuff) [inline internal/services/synapse/synapse_workspace_resource.go:486]: can we add a comment explaining why this is needed, so this isn't removed in the future?

Comment (ms-henglu) [reply internal/services/synapse/synapse_workspace_resource.go]: Sure. And yes.

**Review by katbyte — DISMISSED** [review]
There are a number of failing tests, could we please fix them up?
image

**Comment by ms-henglu** [comment]
Hi @katbyte , I've fixed the failed tests in synapse workspace, and I saw there're more in synapse rp, I think I'll raise another PR to fix them, because they're not related with this one.

**Comment by katbyte** [comment]
@ms-henglu - we can remove all and eny 3.0 flag blocks now. it seems you've made the tests 

```
[](https://ci-oss.hashicorp.engineering/viewLog.html?buildId=270799&buildTypeId=TerraformOpenSource_TerraformProviders_AzureRMPublic_AZURERM_SERVICE_PUBLIC_SYNAPSE#)------- Stdout: -------
=== RUN   TestAccSynapseWorkspaceSecurityAlertPolicy_update
=== PAUSE TestAccSynapseWorkspaceSecurityAlertPolicy_update
=== CONT  TestAccSynapseWorkspaceSecurityAlertPolicy_update
    testcase.go:110: Step 1/4 error: Error running pre-apply refresh: exit status 1
        
        Error: Insufficient identity blocks
        
          on terraform_plugin_test.tf line 26, in resource "azurerm_synapse_workspace" "test":
          26: resource "azurerm_synapse_workspace" "test" {
        
        At least 1 "identity" blocks are required.
--- FAIL: TestAccSynapseWorkspaceSecurityAlertPolicy_update (17.23s)
FAIL
```

**Comment by ms-henglu** [comment]
Hi @katbyte , the failed tests are not introduced by this PR, so I opened another PR to fix the failed synapse testcases.
https://github.com/hashicorp/terraform-provider-azurerm/pull/15899

File: internal/services/synapse/synapse_workspace_resource.go [hunk 2]
```diff
@@ -477,13 +477,15 @@ func resourceSynapseWorkspaceRead(d *pluginsdk.ResourceData, meta interface{}) e
 
 	aadAdmin, err := aadAdminClient.Get(ctx, id.ResourceGroup, id.Name)
 	if err != nil {
-		if !utils.ResponseWasNotFound(aadAdmin.Response) {
+		// AAD admin is not supported in a synapse workspace created from a Dedicated SQL pool (formerly SQL DW)
```
Comment (tombuildsstuff) [inline internal/services/synapse/synapse_workspace_resource.go:480]:
```suggestion
		// NOTE: AAD Admin isn't supported for a Workspace created from a Dedicated SQL Pool / SQL DataWarehouse and returns a Conflict
```

File: internal/services/synapse/synapse_workspace_resource.go [hunk 3]
```diff
@@ -477,13 +477,15 @@ func resourceSynapseWorkspaceRead(d *pluginsdk.ResourceData, meta interface{}) e
...
+		if !utils.ResponseWasNotFound(aadAdmin.Response) && !utils.ResponseWasConflict(aadAdmin.Response) {
 			return fmt.Errorf("retrieving AzureAD Admin for %s: %+v", *id, err)
 		}
 	}
 	sqlAdmin, err := sqlAdminClient.Get(ctx, id.ResourceGroup, id.Name)
 	if err != nil {
-		if !utils.ResponseWasNotFound(sqlAdmin.Response) {
+		// Sql AAD admin is not supported in a synapse workspace created from a Dedicated SQL pool (formerly SQL DW)
```
Comment (tombuildsstuff) [inline internal/services/synapse/synapse_workspace_resource.go:487]:
```suggestion
		// NOTE: SQL Admin isn't supported for a Workspace created from a Dedicated SQL Pool / SQL DataWarehouse and returns a Conflict
```

**Review by tombuildsstuff — APPROVED** [review]
hi @ms-henglu 

Thanks for pushing those changes, I've left a couple of suggestions inline which I hope you don't mind but I'm going to push a commit to fix - but this otherwise LGTM 👍 

Thanks!

**Pull request merged**